handlers/hierarchy.go

```diff
@@ -112,12 +114,11 @@ func (f *Filter) HierarchyUpdate(w http.ResponseWriter, req *http.Request) {
 				}
 			}
 		}
-
-		wg.Done()
 	}()
 
+	options := make([]string, 0)
 	for k := range req.Form {
-		if k == "save-and-return" || k == ":uri" {
+		if k == "save-and-return" || k == ":uri" || k == "q" {
```

Comment:
Being über pedantic it would be more efficient to define these string vars outside of the `for`. As it currently is new vars will be created and assigned these values with each iteration of the loop. A bit of a premature optimisation but the longer this input list is the more inefficient it will be. I will leave it up to you.

Reply:
I don't agree or don't understand.  🤷‍♂ 
Which vars are being created in the loop?

Reply:
I agree with Gedge but you could make the following constants outside the for loop:

`save-and-return`

Reply:
have made that change